testing/unittests/jlibtests.cpp

```diff
@@ -3806,6 +3822,117 @@ class JLibSecretsTest : public CppUnit::TestFixture
+
+        //Sleep so the cache entry should have expired, and no data around to make it not stale.
+        MilliSleep(60);
+        CPPUNIT_ASSERT(!secret4->isValid());
+        CPPUNIT_ASSERT(secret4->isStale());
+
+        //--------- Now update the value in the background ---------
+        //First check that a missed secret is checked in the background
+        Owned<ISyncedPropertyTree> secret5 = getSyncedSecret("testing", "secret5", nullptr, nullptr);
+        CPPUNIT_ASSERT(!secret5->isValid());
+        CPPUNIT_ASSERT(!secret5->isStale());
+        //And write a value so it is picked up on the next refresh
+        writeTestingSecret("secret5", "value", "secret5Value");
+
+        //Sleep for less than the update interval
+        MilliSleep(50); // elapsed=50
+        CPPUNIT_ASSERT(!secret5->isValid());
+        CPPUNIT_ASSERT(!secret5->isStale());
+
+        //Sleep so the cache entry should have expired and the value reread since reading ahead
+        MilliSleep(60); // elapsed=110 = 80 + 30
+        CPPUNIT_ASSERT(secret5->isValid());
+        CPPUNIT_ASSERT(!secret5->isStale());
+
+        //Sleep again so it is not accessed within the timeout period - it should now be marked as stale but valid
+        MilliSleep(100); // elapsed=210 = 80 + 80 + 50
+        CPPUNIT_ASSERT(secret5->isValid());
+        CPPUNIT_ASSERT(secret5->isStale());
+
+        //--------- Check that accessing the function marks the value so it is refreshed ---------
+        Owned<ISyncedPropertyTree> secret6 = getSyncedSecret("testing", "secret6", nullptr, nullptr);
+        CPPUNIT_ASSERT(!secret6->isValid());
+        CPPUNIT_ASSERT(!secret6->isStale());
+        //And write a value so it is picked up on the next refresh
+        writeTestingSecret("secret6", "value", "secret6Value");
+
+        //Sleep for less than the update interval
+        MilliSleep(50); // elapsed=50
+        CPPUNIT_ASSERT(!secret6->isValid());
+        CPPUNIT_ASSERT(!secret6->isStale());
+
+        //Sleep so the cache entry should have expired and the value reread since reading ahead
+        MilliSleep(60); // elapsed=110 = 80 + 30
+        CPPUNIT_ASSERT(secret6->isValid());
+        CPPUNIT_ASSERT(!secret6->isStale());
+        unsigned version1 = secret6->getVersion(); // Mark the value as accessed, but too early to be refreshed
+        writeTestingSecret("secret6", "value", "secret6Value2");
+
+        MilliSleep(40); // elapsed=150 = 80 + 70
+        CPPUNIT_ASSERT(secret6->isValid());
+        CPPUNIT_ASSERT(!secret6->isStale());
+        unsigned version2 = secret6->getVersion(); // Mark the value as accessed, but too early to be refreshed
+        CPPUNIT_ASSERT(version2 == version1);
+
+        MilliSleep(30); // elapsed=180 = 80 + 80 + 20
+        CPPUNIT_ASSERT(secret6->isValid());
+        CPPUNIT_ASSERT(!secret6->isStale());
+        unsigned version3 = secret6->getVersion(); // Mark the value as accessed, but too early to be refreshed
```

Comment:
Is the above comment wrong?  No longer too early to be refreshed?

Reply:
You are correct comment is wrong.